### PR TITLE
[WIP] Make StructureGroups editable from Python

### PR DESCRIPTION
**Pull request by greglandrum**
This adds code to allow StructureGroups to be created and edited from Python.

This may be ready to go as-is, but I'd like to see if there are any comments before merging

This addresses #2489

### CODE REVIEW

File: Code/GraphMol/Wrap/SubstanceGroup.cpp [hunk 1]
```diff
@@ -70,6 +100,48 @@ struct sgroup_wrap {
...
+                SubstanceGroup::setProp<unsigned int>,
+            (python::arg("self"), python::arg("key"), python::arg("val"),
+             python::arg("computed") = false),
+            "sets the value of a particular property")
+        .def("SetBoolProp",
+             (void (RDProps::*)(const std::string &, bool, bool) const) &
+                 SubstanceGroup::setProp<bool>,
+             (python::arg("self"), python::arg("key"), python::arg("val"),
+              python::arg("computed") = false),
+             "sets the value of a particular property")
```
Comment (bp-kelley) [inline Code/GraphMol/Wrap/SubstanceGroup.cpp]: Missing GetPropNames, GetPropsAsDict
```
        .def("GetPropNames", &SubstanceGroup::getPropList,
             (python::arg("self"), python::arg("includePrivate") = false,
              python::arg("includeComputed") = false),
             "Returns a tuple with all property names for this molecule.\n\n"
             "  ARGUMENTS:\n"
             "    - includePrivate: (optional) toggles inclusion of private "
             "properties in the result set.\n"
             "                      Defaults to 0.\n"
             "    - includeComputed: (optional) toggles inclusion of computed "
             "properties in the result set.\n"
             "                      Defaults to 0.\n\n"
             "  RETURNS: a tuple of strings\n")

        .def("GetPropsAsDict", GetPropsAsDict<SubstanceGroup>,
             (python::arg("self"), python::arg("includePrivate") = false,
              python::arg("includeComputed") = false),
             "Returns a dictionary populated with the molecules properties.\n"
             " n.b. Some properties are not able to be converted to python "
             "types.\n\n"
             "  ARGUMENTS:\n"
             "    - includePrivate: (optional) toggles inclusion of private "
             "properties in the result set.\n"
             "                      Defaults to False.\n"
             "    - includeComputed: (optional) toggles inclusion of computed "
             "properties in the result set.\n"
             "                      Defaults to False.\n\n"
             "  RETURNS: a dictionary\n")
```

Comment (bp-kelley) [reply Code/GraphMol/Wrap/SubstanceGroup.cpp]: We should just make an RDProp base class and get rid of this stuff :)

File: Code/GraphMol/Wrap/SubstanceGroup.cpp [hunk 2]
```diff
@@ -105,11 +177,20 @@ struct sgroup_wrap {
...
+                "returns a copy of the molecule's SubstanceGroups (if any)",
                 python::with_custodian_and_ward_postcall<0, 1>());
+    python::def("GetMolSubstanceGroupWithIdx", &getMolSubstanceGroupWithIdx,
+                "returns a particular SubstanceGroup from the molecule",
+                python::return_internal_reference<
+                    1, python::with_custodian_and_ward_postcall<0, 1>>());
     python::def("ClearMolSubstanceGroups", &clearMolSubstanceGroups,
                 "removes all SubstanceGroups from a molecule (if any)");
-    // FIX: needs something tying the lifetime to the mol
+    python::def("CreateMolSubstanceGroup", &createMolSubstanceGroup,
```
Comment (bp-kelley) [inline Code/GraphMol/Wrap/SubstanceGroup.cpp]: This has no corollary in C++ land, correct?  Perhaps there should be one.

File: Code/GraphMol/Wrap/SubstanceGroup.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/Wrap/SubstanceGroup.cpp]: `GetPropNames()` and `GetPropsAsDict()` are already there: https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/Wrap/SubstanceGroup.cpp#L97

Agreed, making an RDProp base class would simplify the wrappers a lot. Definitely something we should do for the next release.

**Comment by bp-kelley** [comment]
Sorry missed that.

File: Code/GraphMol/Wrap/SubstanceGroup.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/Wrap/SubstanceGroup.cpp]: C++ doesn't really need an equivalent. The code there is:
```
  SubstanceGroup sg(&mol, type);
  addSubstanceGroup(mol, sg);
```

**Pull request merged**